### ISSUE TITLE
OpenTelemetry tracing support to extend observability

### ISSUE THREAD

**Issue by saschagrunert** [youki-dev/youki#1348]
I think this project could outstandingly shine when featuring tracing support. We have a implementation in conmon-rs which supports exporting data to the OpenTelemetry exporter (OTLP): https://github.com/containers/conmon-rs/blob/6cfba5a/conmon-rs/server/src/telemetry.rs

Everything is based on the [tracing crate](https://docs.rs/tracing/latest/tracing), which is also used for logging.

What do you think about that?

**Comment by utam0k** [comment]
@YJDoc2 @Furisto @yihuaf I'd like to hear your input.

**Comment by utam0k** [comment]
https://kubernetes.io/blog/2022/12/01/runtime-observability-opentelemetry/

**Comment by saschagrunert** [comment]
@alban mentioned in https://social.linux.pizza/@albcr/109437558144011212 that we could use an annotation to pass the trace ID around. This way we can link youki/runc/crun with higher level runtimes like CRI-O.

**Comment by YJDoc2** [comment]
Hey, I haven't yet looked into this in depth, but from what I have seen, and read this seems like a good idea if we can support. We might not need to make too many changes for it, however I think we must feature-gate this, and disable it by default. As stated in the blog, this will cause some performance degradation, and given that _usually_ a user will need some additional tools to take full advantage of the spans and stuff, it is better left to the user to decide if to opt-in or not.

I'll try taking a more detailed look and maybe add details on how we could go on adding the support.

**Comment by lengrongfu** [comment]
> @alban mentioned in https://social.linux.pizza/@albcr/109437558144011212 that we could use an annotation to pass the trace ID around. This way we can link youki/runc/crun with higher level runtimes like CRI-O.

For low-level runtimes such as `youki` or `runc`, they are more called by `containerd` high-level runtime. I think we should first communicate with upstream `containerd` on how to transmit traceID and spandID.

**Comment by yihuaf** [comment]
Introducing `tracing` to `youki` would be a great idea. We can also clean up the logging libs we have, since logging becomes a special case of tracing. This seems to be where the rust ecosystem is moving to anyway, so we should embrace it. I will take a jab at this after I migrate the libcontainer to `thiserror`.

**Comment by utam0k** [comment]
@yihuaf 👍

**Comment by utam0k** [comment]
@saschagrunert Do you have an idea how to get the tracing id from the CRI layer? container id?

**Comment by utam0k** [comment]
@yihuaf I have implemented a trial for this. And I have lost that code 😭 
But I remember the most difficult point. That is how to follow the trace when double-forking. I ran into this problem and left this issue for a time. I am sharing it once.

**Comment by yihuaf** [comment]
@utam0k I got a prototype roughly working except the opentelemetry related crates really wants `async`. I don't think there is a way to use opentelemetry without an `async` runtime. I am not sure about all the implication  of introducting an async runtime into our codebase yet. Did you run into this issue previously?

**Comment by yihuaf** [comment]
> First of all, it might be good to have an option by features or arguments.

I agree we should make this into a feature flag. I will further verify the correctness of the setup across the process boundary. At minimum, looks like this observability feature is best served as a "debugging mode".

**Comment by utam0k** [comment]
> At minimum, looks like this observability feature is best served as a "debugging mode".

💯 
I may do it, but I think that is a good idea until standardization is achieved.

**Comment by saschagrunert** [comment]
Awesome to read, thank you so much for pushing the topic forward folks!

**Comment by Mossaka** [comment]
Hey @utam0k and @yihuaf, 

I am excited to see youki is moving forward with the OpenTelemetry implementation. It would be great to integrate youki tracing alongside Runwasi tracing spans at the containerd's OpenTelemetry endpoint. More details on the containerd issue can be found here: (https://github.com/containerd/containerd/issues/10173)

One of the core issues the community needs to figure out is establishing a method  to pass tracing context between containerd and the shim, ensuring correlation between containerd ttrpc requests and shim responses. 

From my experience adding OpenTelemetry tracing support to Runwasi https://github.com/containerd/runwasi/issues/10, I've gained some insights which I think would be valuable to the discussion here.

> But I remember the most difficult point. That is how to follow the trace when double-forking. I ran into this problem and left this issue for a time.

I ran into the same issue when I first started implementing tracing for runwasi. Eventually I figured that the parent process needs to serialize the [OpenTelemetry Context](https://opentelemetry.io/docs/concepts/context-propagation/) and pass that to the child process, where it sets it's current Span to the parent's context. (See https://github.com/containerd/rust-extensions/issues/268)

> I don't think there is a way to use opentelemetry without an async runtime.

I agree it's necessary at this stage. 

> but I think that is a good idea until standardization is achieved.

As for the topic of standardization, what specific aspects do you think need addressing? I am happy to push this forward and find a solution that facilitates seamless end-to-end tracing from containerd to shims and to runtimes like youki.

Additionally, are there any ongoing efforts to enhence `tracing` support beyond what has been implemented for logging https://github.com/containers/youki/pull/1899?
